### ISSUE TITLE
MySQLProfile has a significant change in UpsertBuilder logic between 3.3.3 and 3.4.1

### ISSUE THREAD

**Issue by MarmaladeSky** [slick/slick#2784]
**DESCRIPTION:**
As it is mentioned in the title, MySQLProfile has a huge change in `insertOrUpdate` logic between 3.3.3 and 3.4.1. 
TL;RD: in `3.3.3`, primary key column is not included in the `on duplicate key update` list, but it is included in `3.4.1`.  
This cause primary key "jump" if update is made by duplicate on another unique column.

**STEPS TO REPRODUCE:**
```scala
import slick.jdbc.JdbcBackend
import slick.jdbc.MySQLProfile.api._

import scala.concurrent.ExecutionContext.Implicits.global
import scala.concurrent.duration._
import scala.concurrent.{Await, Future}

object MysqlUpsertBug {

  private val db: JdbcBackend#DatabaseDef = Database.forConfig("mysql")

  case class AnimalRow(id: Int, name: String, location: String)
  class Animals(tag: Tag) extends Table[AnimalRow](tag, "animals") {
    def id = column[Int]("id", O.PrimaryKey, O.AutoInc)
    def name = column[String]("name", O.Unique, O.Length(100))
    def location = column[String]("location")
    def * = (id, name, location) <> (AnimalRow.tupled, AnimalRow.unapply)
  }
  val animals = TableQuery[Animals]

  private def init(): Future[Unit] = {
    db.run {
      DBIO.seq(
        animals.schema.dropIfExists,
        animals.schema.create
      )
    }
  }

  def main(args: Array[String]): Unit = {
    val f = for {
      _ <- init()

      tigerRow = AnimalRow(0, "tiger", "NY")
      _ <- db.run { animals.insertOrUpdate(tigerRow) }
      tigerAfterInsert <- db.run { animals.filter(_.name === "tiger").result.head }

      movedTiger = tigerRow.copy(location = "CA")
      _ <- db.run { animals.insertOrUpdate(movedTiger) }
      tigerAfterUpsertByUniqueName <- db.run { animals.filter(_.name === "tiger").result.head }

    } yield assert(tigerAfterInsert.id == tigerAfterUpsertByUniqueName.id)

    Await.result(f, Duration.Inf)
  }

}

```

**EXPECTED RESULT:**
The code above is executed without assertion failure. After the execution, DB contains a single row with id = 1.

**ACTUAL RESULT:**
In 3.3.3 the expected result is provided, in 3.4.1 the assertion failed, the DB contains a single row with id = 2.
`3.3.3`
```
insert ignore into `animals` (`id`,`name`,`location`)  values (0,'tiger','CA') on duplicate key update `name`=VALUES(`name`), `location`=VALUES(`location`)
```
`3.4.1`:
```
insert into `animals` (`id`,`name`,`location`)  values (0,'tiger','NY') on duplicate key update `id`=VALUES(`id`), `name`=VALUES(`name`), `location`=VALUES(`location`)
```

**SETUP:**
All the setup and easy to reproduce way is available [here](https://github.com/MarmaladeSky/slick_mysql_upsert_bug) (see launch.sh).
There are two branches, `slick_3.3.3` and `slick_3.4.1`.

**Comment by nafg** [comment]
@jkugiya can you help?

**Comment by jkugiya** [comment]
@MarmaladeSky @nafg I need your feedback.
https://github.com/slick/slick/issues/2784#issuecomment-1643935013

**Comment by MarmaladeSky** [comment]
Hello @jkugiya, thank you for your patience.

TL;DR: My feedback is yes, this behavior is expected.


As I understand, in the context of a **single** unique constraint (the primary key in these examples):
- if a part of the unique constraint is AutoInc, `insertOrUpdate` always insert
- if no part of the unique constraint is AutoInc, `insertOrUpdate` may insert or delete based on values provided

I don't see any other options for this behavior. Anything else is going to be much more confusing.

**BUT** 
I more care about consistent behavior unless we're not in the `Slick 3 ~> Slick 4` change set.
Yes, I know that Slick is not on SemVer, but I wish a good luck to anybody upgrading `3.3.3` (the latest for a long time) to `3.4.1`. Because this person is going to have a fun time finding why his PKs jumping if his `insertOrUpdate` is not covered with testing.

I [implemented the above behavior line by line](https://github.com/MarmaladeSky/slick_mysql_upsert_bug/blob/slick_3.3.3_spec/src/main/scala/com/github/marmaladesky/MysqlUpsertBug.scala) and checked both `3.3.3` and `3.4.1`.
Both behave the same way as it is described.

**Comment by jkugiya** [comment]
Of course, any missing tests will be added with modifications.

**Comment by nafg** [comment]
@MarmaladeSky Do we need a 3.4.2 or can the fix just go into the next milestone of 3.5.0?

**Comment by nafg** [comment]
@MarmaladeSky @jkugiya my brain is fried, if either of you could do a call on Discord or whatever it will help me process this (not to mention it will help me carve out time for it)

**Comment by jkugiya** [comment]
#2785

**Comment by jkugiya** [comment]
https://github.com/jkugiya/slick/tree/try/test_3.3.3

https://github.com/jkugiya/slick/blob/try/test_3.3.3/slick-testkit/src/main/scala/com/typesafe/slick/testkit/tests/InsertTest.scala#L207-L316

The same test was performed with version 3.3.3, and it was confirmed that it works fine here as well.

**Comment by MarmaladeSky** [comment]
@nafg 
>@MarmaladeSky Do we need a 3.4.2 or can the fix just go into the next milestone of 3.5.0?  

Indeed, we need `3.4.2`. Considering that Slick is the main and, I would argue, the default high level DB access library of Scala, I believe such a serious change in behavior should be addressed everywhere it is presented.

>if either of you could do a call on Discord or whatever it will help me process this 

I am available to assist if necessary. However, I prefer to avoid using proprietary, non-indexed chat or video communication platforms in the context of free software projects in order to maintain transparency. If you need PR, test coverage or anything concrete from my side, please feel free to ask.

Just a reminder, the whole issue reproduction is available [here](https://github.com/MarmaladeSky/slick_mysql_upsert_bug) at branches `slick_3.3.3` and `slick_3.4.1`.

**Comment by MarmaladeSky** [comment]
@jkugiya @nafg 
If I can somehow help with the issue or release process, please let me know.

**Comment by jkugiya** [comment]
https://github.com/slick/slick/pull/2785#issue-181406905

Do you feel comfortable with what is written or what changes have been made?
@nafg @MarmaladeSky

**Comment by nafg** [comment]
I'm very overwhelmed by work right now, if someone can have a call that would enable me to focus on this

**Comment by PigerCattus** [comment]
>if either of you could do a call on Discord or whatever it will help me process this
>if someone can have a call that would enable me to focus on this
@nafg 
I'm sorry, I don't get the process. What kind of call, and what do you expect there?

**Comment by MarmaladeSky** [comment]
>if either of you could do a call on Discord or whatever it will help me process this
>if someone can have a call that would enable me to focus on this

@nafg @jkugiya  
I'm sorry, but I don't get the process.
As I see, there are
- [bug report](https://github.com/slick/slick/issues/2784) with as many details as possible
- [PR](https://github.com/slick/slick/pull/2785) that partially takes bug report for tests implementation

What are the next steps?

**Comment by nafg** [comment]
> > if either of you could do a call on Discord or whatever it will help me process this
> > if someone can have a call that would enable me to focus on this
> 
> @nafg @jkugiya I'm sorry, but I don't get the process. 

It's not matter of process. Can you do a call or not?

**Comment by jkugiya** [comment]
It is difficult to take time out, so please verbalise a little more about what is unclear. It is not necessarily a problem that our voices are not heard, is it?

**Comment by MarmaladeSky** [comment]
Hello @nafg @jkugiya,
Apologies for the delay. I wanted to provide an update on this thread.

Last week, I had a short call with @nafg about this issue, thank you for your efforts and a few hints about slick testing (it is a bit tricky). 

During the call, @nafg had sowed doubt into my mind about how the upsert should be processed, so, I decided to take time and try to represent my thoughts with the PR.

All my doubts had gone away after I just put a silly test about a traveling tiger into `InsertTest.scala` of `testkit` submodule.

The test is performed at all DBs with the `insertOrUpdate` capacity [`sqlite`, `postgresql`, `sqlserver`, `oracle`, `hsqldb`, `h2`, `mysql`, `derby`, `db2`] and the only failure is at MySQL profile.

From my perspective, the current PR is ready to be merged and released.

**Comment by MarmaladeSky** [comment]
Should we close this one?

**Comment by nafg** [comment]
@MarmaladeSky IIRC https://github.com/slick/slick/pull/2785 fixed it for 3.5.x but you wanted it backported, so @jkugiya opened https://github.com/slick/slick/pull/2814 which was waiting for you to test it and not sure what happened after that.

But if you're good feel free to close it, and let me know if #2814 should be closed.
